---
title: Lab assignment-1
format:
    html:
        embed-resources: true
---

Read and work through all code in the lab demonstration section, and do all exercises below

**Note**: The version on the website is just an HTML preview, please go to the share-point folder to download the actual assignment and dependencies.

**SUBMISSION**: Upload the completed *assignment* to Canvas (in HTML or PDF). Make sure ALL images & outputs are included! Please do not submit the completed demonstration, only the completed assignment.


## Problem-1

* This relates to Example 1 in Lab-1, you can recycle code from there 
* Mike had the first three successes in trials 6,8 , and 9 . He had six failures until he reached three successes. Do you think Mike has success probability $p=0.5$ or better? Can a simulation give an answer? To do this do.
  * simulate the number of tosses needed to get to three successes, use the probability success probability p = 0.5. (Hint: use the “mytoss” function we did in the Lab class)
  * Run many simulations (say 10,000) with this success probability. (use the “myattempts” function from the class)
  * If Mike’s success probability were 0.5 or better, he would not need a lot of attempts. Find the fraction of simulations where three successes were reached after 9 tosses or later by somebody with success probability 0.5.


In [3]:
mytoss = function(p){
  u = runif(1)
  x = as.numeric(u < p)
  return(x)
}

In [4]:
# This function calculates the number of total attempts to get a success

myattempts = function(p){
  counter = 1
  while (mytoss(p) == 0){ counter = counter + 1 }
  return(counter) 
}

In [5]:
replicate(10,mytoss(.5))

[1] 0 0 1 1 0 0 0 1 0 1

In [6]:
set.seed(23)
replicate(3,myattempts(0.5))

[1] 2 1 3

This means there were 2 attempts to get a first scuccess, 1 attempt to get a second success, and 3 attempts to get a third success, which is total of 6 attempts.


In [7]:
# Number of total attempts until the 3 successes

sum(replicate(3,myattempts(0.5)))

[1] 10

In [8]:
# 10000 simulations of calculating the total number of attempts until the 3 successes

replicate(10000,sum(replicate(3,myattempts(0.5))))

[1]  9  8  4  6  6  7  6  4  9  7  7  3  7  4  7  8  6  6  3  6  4  5  9  7
   [25]  9  5 10  3 14  6  6  7  7  7  3  6  4  9 12  5 11  5  5  7  3  5  6  6
   [49]  5  7  6 11  4  3  8 11  3  4 11  4  7  3  3  6  4  3  5  5  5  5  3  3
   [73]  5  5  4  9  8  6  4  7  5  7  3  5  4  8  7  4  4 10  9  8  4  6  5  4
   [97]  6  6  5  9  6  7  9 10  4  8  5  8  6 10  5  6  3 11  5  9  7  8  7  6
  [121]  6  9  4  9  8  8  4  6  9  6  3  5  5  8  4  8  9  7  4  4  3  6  5  5
  [145]  4  6  4 12  6  3  9  7  3  4  6  6  3  6  7  4  5  8  5  3 10  5 10  6
  [169] 12  6  8  9  6  7 10  8  7  5  3  9  4  7  4  4  6  8  6  9  4  7 10  6
  [193]  4  5  8  5  4  7 10  5  5  3 10  3  9 19  5  4  6  6  3  7  7  4  4  4
  [217]  5  4  5  3  3  5  3  6  4  3  3  6 10  5  5  5  6  7  5 12  6 11 10  9
  [241]  5  6  5  6  7  8  7  4  6  5  5  9  9 12  6  3  5  4  6  6  3  6  6  4
  [265]  5  4  4  8  4  5  9  6  7  8  6  8  8  8  3 10  3  5  7  8  8  8  5  6
  [289]  4  7  5  8  4  7  8  4  6  5  5  3  3  3  4  7  7  8  6  5  9  4  6  9
  [313]  5  6  4  5  8  6  6  5  5  3  4  6  4  5  6  6  5 15  5  3  3  6 12  6
  [337]  8  5 10 12  4  6  6  5  3  5  7  6  4  7  3  5  7  8  5 11  7  5 14  5
  [361]  4  5  5  6  8 10  3  6  5  7  7  5  6  6  6  4  5 12  5  3  7  7  4  7
  [385]  3  6  9  5  3  3  5  6  6  8  4  5  7  3  4  4  3  8  4  5  3  9  4  6
  [409]  6  5  6  6  3  4  7  4  5  8  3  5  4  6  3  3  4  4  6  8 10  9  7  4
  [433]  3 13  6  5  3  8  4  5  4  7  6  3  5 10 10  7  3  4  3 11 10  5  5  3
  [457]  5 11  5  4  3 10  4  3  4 11  5  5  4  5  4  7  4  5  6  3 11  6  3  8
  [481]  6  9  9 10  5  5 11  4  6  5 11  3  7  5  8 11  5  4  5  4  4  5  6  5
  [505]  6  4 13  4  4  9  6  9  8  3  8  5  7  7  4  6  9  5  6  7  5  5  4  9
  [529]  4  8  6  3  5  6  4  4  5  3  8  4  9  5  6  6  4  3  6  7  9  5  8  6
  [553]  7  4  8  3  5  9  4  9  4  8  3  5  3  6  6  6  4 10  6 10  6  4  4  6
  [577]  5  4  5  8  3  4  6  6  4  8  7  4  6  4  4  8  6  5  6  4  3  7  7  3
  [601]  7  5  6  6  4 10  6  9  7  6  7  4  7  4  8  6  7  4  4  4  3  7  4  7
  [625]  5  6  6 10  3  3  3  5  4  7  7 13  5  7  8  4  4  4  4  8  5  4  5  5
  [649]  8  4  7  4  5  3  5 11  8  6  5 16  7  7  6  5  6  5  6  3  5  4  3  5
  [673]  8  3  6  8 12  8  5  6  6  7  7  7  4  8  5  8  4  6  4  4  4  9  9  3
  [697]  6  4  7  7  3  6  7  5  4  5  4  6  3  6  7  5  5 12  3  7  4  5  6  8
  [721] 12 11  4  5  6  4  6  8  5  6  4  6  5  4  5  5  9  4  5  7  8  4  3  8
  [745]  6  3 11  3  9  3  5 10  8  4  5  6  6  8 10  3  6  5  4  6  6 14  5  6
  [769]  8  4  5  3  7  7  5 10  5 11  5  3  3  7  5  7 10  3  5  4  3  9  7  7
  [793]  5  7  5  7  6  7  7 11  3  4 15  3  7  6  5 12  6  6  7  4  4  6  6  5
  [817]  6 12 10  6  3  8  8  3  7  6 11  7  6  9  3 12  5  9  6  9  5  8  3  8
  [841]  7  8  3  3 11  4  4  8  3  6  9  4  8  3  7  6  5  5  3  5  6  5  4  5
  [865]  6  3  4  7  4  8 13  3  6  5 10  3  5  8  6 10 10  3  8  9  6  5  7  7
  [889]  8 14  6  6 11  7  3  9  5  9  8  5  3  7  4 11  5  5  3  5  8  4  9  3
  [913] 10  9  6  6  3  6  5  4  4  3  6 11 10  6  6  4  3  3  9  5  3  3  4  8
  [937]  5  5  4  6  4  8  4  5  5  4  8  3  7  5  4  5  4  8 10  3  3  4  5  4
  [961]  9  4  8  7  4  4  7  9  6  4  6  6  6  9  7  6  4  8  6  6  5  9  3  5
  [985] 11  4  5  5  8  9  4  3  6  7  7 12  6  9  6  4  6 21  3  5  8  4  6  5
 [1009]  4  6  9  5  3  5  3  5  8  6  8  6  3  4  4  5  7 10  5  7  4  5  6  5
 [1033]  5  4  6  6 12  8  8  4  9  4  9  3  3  5  4  7  6  8  7  6  7  7  5 15
 [1057]  3  6  5 10  8  8  5  7  3  8 13  5 10  3  3  9  3  5  5  3  6  4  5 12
 [1081]  5  7 12  6  5 10  4  3  9  4  4  6  3  7  4  4 13  9  5  9  4  6  5  6
 [1105]  3  5  5  4  5  3  5  8  8  7  7  5 12  9  9  5  3  8  6  4  6  6  3  7
 [1129]  8  4  8  4  3 11 11  4  6  6  5  5  8  3  4  8  5  8  3  3  3  4  9  3
 [1153]  8  5  5  5  7  9  3  4  6  7  6  7  5  6 12  7  9  4  6  5  6  5  5  3
 [1177]  7  8  6  4  7  7 12  6  3  5  5  4  7  3  3  7  3  4  9  5  4  4 11  7
 [12

In [9]:
# Number of simulations where Mike needs more than 9 attempts to get the 3 successes

s = sum(replicate(10000,sum(replicate(3,myattempts(0.5))))>=9)
s
s/10000 # Calculating the fraction

[1] 1391

[1] 0.1391

In [10]:
myattempts2 = function(p){
  counter = 0
  while (mytoss(p) == 0){ counter = counter + 1 }
  return(counter) 
}

sum(replicate(10000,sum(replicate(3,myattempts2(0.5))))>=6)/10000

[1] 0.1464

The fraction of simulation where it needs more than 9 attempts to get the 3 successes is 0.0925 when the success probability is 0.5. This means there is 90.75% chance that this person gets 3 successes in less than or equal to 9 attempts. It implies, therefore, Mike is most likely to have a better probability than 0.5, because he got 3 successes in 9 attempts.

## Problem 2: 

Baby names for male and female babies.

Repeat Example 2 using `data/yob2010.txt`. Interpret your results (in wording relates to the problem) for each simulation step by step.

Do the following 

* Read csv file with baby names and make new column names
* Make a the table function returns a vector whose components have names.
* Compute the number of female and male births, using the aggregate() function
* Sample 30 names with replacement. Use probabilities that are proportional to the counts.
* Compute actual probabilities of occurrence
* Find all names which occur both as male and female baby names. First make subsets of female and male births. Then find the set intersection of the names.
* Find out how often each of these names is used for females and males. Use sub-setting to make data frames of female and male birth data for names that are used for both genders. We may drop the gender variable. 
* Now merge the two data frame by names. R finds the merge variable (i.e. name) automatically.


In [8]:
# Read csv file with baby names and make new column names

df = read.csv("data/yob2010.txt", header = FALSE)
names(df) = c("names","gender","count")
head(df)

names    gender count
1 Isabella F      22883
2 Sophia   F      20612
3 Emma     F      17322
4 Olivia   F      17012
5 Ava      F      15418
6 Emily    F      14260

In [9]:
# Make a the table function returns a vector whose components have names.

table(df$gender)


    F     M 
19800 14241 

There are 19800 female names and 14241 male names.


In [10]:
# Compute the number of female and male births, using the aggregate() function

aggregate(count ~ gender, data = df, FUN = "sum")

gender count  
1 F      1772738
2 M      1913851

There are 1772738 births of females and 1913851 births of males.


In [11]:
# Sample 30 names with replacement. Use probabilities that are proportional to the counts.
# Compute actual probabilities of occurrence

df$probs = df$count/sum(df$count)
head(df)
sample(df$names,30,replace=TRUE,prob=df$probs)

names    gender count probs      
1 Isabella F      22883 0.006207093
2 Sophia   F      20612 0.005591076
3 Emma     F      17322 0.004698652
4 Olivia   F      17012 0.004614564
5 Ava      F      15418 0.004182186
6 Emily    F      14260 0.003868074

[1] "Dayanna"   "Ruben"     "Bridger"   "Anthony"   "Tamia"     "Jadyn"    
 [7] "Richard"   "Nevaeh"    "Nicholas"  "Liberty"   "Joaquin"   "Curtis"   
[13] "Iker"      "Finn"      "Stacy"     "Daisy"     "Andre"     "Jayden"   
[19] "Sienna"    "Lakayla"   "Sophia"    "Luke"      "Brett"     "Gavin"    
[25] "Paxton"    "Andrea"    "Cedric"    "Harper"    "Mackenzie" "Jose"

Computing the probabilities of occurrence of each name, it is possible to sample 30 names using those probability.


In [12]:
# Find all names which occur both as male and female baby names. First make subsets of female and male births. Then find the set intersection of the names.

femalebirths = df[df$gender == "F",]
malebirths = df[df$gender == "M",]
both = intersect(femalebirths$name, malebirths$name)
head(both)
length(unique(both))

[1] "Isabella" "Sophia"   "Emma"     "Olivia"   "Ava"      "Emily"

[1] 2438

There are 2438 names used on both female and male babies.


In [13]:
# Find out how often each of these names is used for females and males. Use sub-setting to make data frames of female and male birth data for names 
# that are used for both genders. We may drop the gender variable.

df.female.both <- df[is.element(df$name,both) & df$gender == "F",c(1,3)]
names(df.female.both) <- c("name","female.count")
head(df.female.both)
df.male.both <- df[is.element(df$name,both) & df$gender == "M",c(1,3)]
names(df.male.both) <- c("name","male.count")
head(df.male.both)

name     female.count
1 Isabella 22883       
2 Sophia   20612       
3 Emma     17322       
4 Olivia   17012       
5 Ava      15418       
6 Emily    14260

name      male.count
19801 Jacob     22082     
19802 Ethan     17985     
19803 Michael   17308     
19804 Jayden    17152     
19805 William   17030     
19806 Alexander 16742

Among the names used in both gender, Isabella, Sophia, Emma, Olivia, Ava, and Emily are used the most in female babies, and Jacob, Ethan, Michael, Jayden, William, and Alexander are used the most in male babies.


In [14]:
# Now merge the two data frame by names. R finds the merge variable (i.e. name) automatically.

df.both = merge(df.female.both,df.male.both)
head(df.both)

name    female.count male.count
1 Aaliyah 4657            6      
2 Aamari     8            8      
3 Aaren      5           31      
4 Aarin      7           21      
5 Aaron     23         7450      
6 Aarya     72           26

## Problem-3

* Problem 1.5 in ch. 1 of Dalgaard. On p. 27, replicate was used to simulate the distribution of the mean of 20 random numbers from the exponential distribution by repeating the operation 10 times. That code is `replicate(10,mean(rexp(20)))`

How would you do the same thing with sapply?


In [15]:
replicate(10,mean(rexp(20)))

[1] 1.0361228 0.9232609 1.1972733 1.1732102 1.1080650 1.5683455 1.3951982
 [8] 0.7829826 1.2815841 0.9695471

In [16]:
sapply(data.frame(replicate(10,rexp(20))),mean)

X1        X2        X3        X4        X5        X6        X7        X8 
1.0286909 1.2703632 0.9991616 1.1264687 1.0137392 0.9397388 0.9597289 1.2411207 
       X9       X10 
0.8178500 0.7649052